api/src/main/java/org/reactivestreamsio/tcp/TcpConnection.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * <h2>Flush</h2>
+ *
+ * In order to optimize the number of system calls to write data on the underneath socket, this connection breaks the
+ * writing process into two parts, viz., write and flush. A flush call writes all previous writes on this connection on
+ * the underneath socket.
+ *
+ * <h2>Write errors</h2>
+ *
+ * Every write call returns a {@link Publisher} that represents the acknowledgment of the write action. If the write
+ * failed, then the subscriber's {@link Subscriber#onError(Throwable)} method will be called and if it succeeds, then
+ * subscriber's {@link Subscriber#onComplete()} method will be called
+ *
+ * <h2>Flush errors</h2>
+ *
+ * Since, a {@link #flush()} is a batch operation encompassing all previous unflushed writes, it will represent the
+ * result of all those writes. It will fail if any one of the writes fail and it will not be possible to correlate which
+ * write caused the flush to fail. In order to get granular result per write, one has to subscribe to the
+ * {@link Publisher} of that particular write.
+ *
+ * <h2>Close</h2>
+ *
+ * Since, this connection only allows a single subscription, cancelling that subscription
+ * (via {@link Subscription#cancel()}) closes the connection.
+ * For cases which ignores content (via {@link #ignoreContent()}), an explicit {@link #close()} is required to be
+ * invoked for closing this connection.
+ *
+ * @param <R> Type of objects read from this connection.
+ * @param <W> Type of objects written to this connection.
+ */
+public interface TcpConnection<R, W> extends Publisher<R> {
+
+    /**
+     * Subscribes to the content stream of this connection.
+     *
+     * <h2>Multiple subscriptions</h2>
+     * This connection only allows a single subscription to the content stream. Any subsequent subscription will result
+     * in invocation of {@link Subscriber#onError(Throwable)} on that subscriber.
+     *
+     * @param subscriber Subscriber for the content.
+     */
+    @Override
+    void subscribe(Subscriber<? super R> subscriber);
+
+    /**
+     * Ignores all content of this connection. This is equivalent to calling {@link #subscribe(Subscriber)} with a
+     * subscriber that discards all received items.
+     */
+    void ignoreContent();
+
+    /**
+     * On subscription of the returned {@link Publisher}, writes the passed {@code msg} on this connection.
+     * Every subscription will write the same {@code msg} on the connection.
+     *
+     * <b>This does not flush the write.</b> Call {@link #flush()} to flush this write.
+     * @param msg Message to write.
+     *
+     * @return A {@link Publisher} representing the result of this write. One has to subscribe to this {@link Publisher}
+     * to trigger the write. Each subscription will trigger a write of the same message. The subscriber to this
+     * {@link Publisher} will not complete unless {@link #flush()} is called on this connection.
+     */
+    Publisher<Void> write(W msg);
```

Comment:
It is to enable composition. If it was just a callback then RS libraries would immediately have to convert the callback into an RS to compose, such as with `flatMap` in order to wait on a write or handle errors. It would be a `Publisher<Void>` since it is just the terminal `onError`/`onComplete` states that matter for a write.